File: src/shared/stats.ts
// Helpers de consulta sobre o catálogo de bônus/atributos de itens (D4; dado puro em statData.ts).
// Base do U11 (filtro/seleção por status) e do futuro modelo de stats (H10/H11).
import {
  AFFIX_REP,
  GRADE_SLOTS,
  MOD_TYPES,
  STAT_IDS,
  STAT_MODS,
  STAT_STRINGS,
  type GradeSlots,
  type StatMod
} from './statData'

export type StatId = (typeof STAT_IDS)[number]
export type ModType = (typeof MOD_TYPES)[number]

// Um tipo de status para listas de seleção (U11): id canônico + rótulo + linha-template PT-BR.
export interface StatType {
  id: string
  name: string
  line: string
}

/** Rótulo curto PT-BR do status (ex.: "Armadura"); cai para o próprio id se desconhecido. */
export function statName(id: string): string {
  return STAT_STRINGS[id]?.name ?? id
}

/** Linha-template do status (ex.: "Armadura +{0}"); cai para "<id> +{0}". */
export function statLine(id: string): string {
  return STAT_STRINGS[id]?.line ?? `${id} +{0}`
}

/** Formata um número no padrão pt-BR (inteiro quando possível, senão até 2 casas). */
function fmtValue(value: number): string {
  return new Intl.NumberFormat('pt-BR', {
    maximumFractionDigits: Number.isInteger(value) ? 0 : 2
  }).format(value)
}

/**
 * Preenche a linha do status com um valor (ex.: `formatStatLine('Armor', 35)` → "Armadura +35").
 * Substitui o template `{0}` da linha; se não houver `{0}`, anexa "+valor".
 */
export function formatStatLine(id: string, value: number): string {
  const line = statLine(id)
  const v = fmtValue(value)
  return line.includes('{0}') ? line.replace('{0}', v) : `${line} +${v}`
}

// Lista de status para seleção (U11), ordenada pelo rótulo PT-BR.
export const STAT_LIST: StatType[] = STAT_IDS.map((id) => ({
  id,
  name: statName(id),
  line: statLine(id)
})).sort((a, b) => a.name.localeCompare(b.name, 'pt-BR'))

// Um modificador rolável já resolvido (legível), derivado de STAT_MODS.
export interface ResolvedMod {
  key: string // "baseKey:level"
  statId: string
  modType: ModType
  min: number
  max: number
}

function resolveMod(key: string, mod: StatMod): ResolvedMod {
  const [statIdx, mtIdx, min, max] = mod
  return {
    key,
    statId: STAT_IDS[statIdx] ?? '?',
    modType: MOD_TYPES[mtIdx] ?? MOD_TYPES[0],
    min,
    max
  }
}

/** Todos os modificadores rolaveis de um status (ranges min/max por nível), ordenados por min. */
export function modsForStat(statId: string): ResolvedMod[] {
  const out: ResolvedMod[] = []
  for (const key of Object.keys(STAT_MODS)) {
    const mod = STAT_MODS[key]
    if (STAT_IDS[mod[0]] === statId) out.push(resolveMod(key, mod))
  }
  return out.sort((a, b) => a.min - b.min || a.max - b.max)
}

/** Faixa global (min/max) que um status pode rolar entre todos os seus modificadores. */
export function statRange(statId: string): { min: number; max: number } | null {
  const mods = modsForStat(statId)
  if (mods.length === 0) return null
  return {
    min: Math.min(...mods.map((m) => m.min)),
    max: Math.max(...mods.map((m) => m.max))
  }
}

/** Slots de afixo de uma raridade (quantos bônus o item comporta). */
export function gradeSlotsFor(grade: string): GradeSlots | null {
  return GRADE_SLOTS[grade] ?? null
}

/** Total de slots de afixo de uma raridade (soma dos tipos de slot). */
export function gradeSlotTotal(grade: string): number {
  const s = GRADE_SLOTS[grade]
  if (!s) return 0
  return s.inherent + s.deco + s.engr + s.inscr + s.extra
}

/** Representação/orçamento de afixo de um status (ou null se não catalogado). */
export function affixRep(statId: string): (typeof AFFIX_REP)[string] | null {
  return AFFIX_REP[statId] ?? null
}

File: dotnet/TbhTracker.Core/Logic/Stats.cs
using System.Globalization;
using TbhTracker.Core.Models;

namespace TbhTracker.Core.Logic;

// Port de src/shared/stats.ts (catalogo cru vem de statData.json).

public sealed class StatType
{
    public string Id { get; init; } = "";
    public string Name { get; init; } = "";
    public string Line { get; init; } = "";
}

public sealed class ResolvedMod
{
    public string Key { get; init; } = "";
    public string StatId { get; init; } = "";
    public string ModType { get; init; } = "";
    public double Min { get; init; }
    public double Max { get; init; }
}

public static class Stats
{
    private static readonly CultureInfo PtBr = CultureInfo.GetCultureInfo("pt-BR");
    private static StatDataModel D => Catalog.Stats;

    public static string StatName(string id) =>
        D.StatStrings.TryGetValue(id, out var s) ? s.Name : id;

    public static string StatLine(string id) =>
        D.StatStrings.TryGetValue(id, out var s) ? s.Line : $"{id} +{{0}}";

    /// <summary>Numero no padrao pt-BR (inteiro quando possivel, senao ate 2 casas).</summary>
    public static string FmtValue(double value)
    {
        var isInt = value == Math.Floor(value);
        return value.ToString(isInt ? "#,##0" : "#,##0.##", PtBr);
    }

    public static string FormatStatLine(string id, double value)
    {
        var line = StatLine(id);
        var v = FmtValue(value);
        return line.Contains("{0}") ? line.Replace("{0}", v) : $"{line} +{v}";
    }

    private static List<StatType>? _statList;
    public static IReadOnlyList<StatType> StatList => _statList ??= D.StatIds
        .Select(id => new StatType { Id = id, Name = StatName(id), Line = StatLine(id) })
        .OrderBy(s => s.Name, StringComparer.Create(PtBr, false))
        .ToList();

    private static ResolvedMod ResolveMod(string key, double[] mod)
    {
        var statIdx = (int)mod[0];
        var mtIdx = (int)mod[1];
        return new ResolvedMod
        {
            Key = key,
            StatId = statIdx >= 0 && statIdx < D.StatIds.Count ? D.StatIds[statIdx] : "?",
            ModType = mtIdx >= 0 && mtIdx < D.ModTypes.Count ? D.ModTypes[mtIdx] : D.ModTypes[0],
            Min = mod[2],
            Max = mod[3]
        };
    }

    public static List<ResolvedMod> ModsForStat(string statId)
    {
        var outList = new List<ResolvedMod>();
        foreach (var (key, mod) in D.StatMods)
        {
            var idx = (int)mod[0];
            if (idx >= 0 && idx < D.StatIds.Count && D.StatIds[idx] == statId)
                outList.Add(ResolveMod(key, mod));
        }
        return outList.OrderBy(a => a.Min).ThenBy(a => a.Max).ToList();
    }

    public static (double Min, double Max)? StatRange(string statId)
    {
        var mods = ModsForStat(statId);
        if (mods.Count == 0) return null;
        return (mods.Min(m => m.Min), mods.Max(m => m.Max));
    }

    public static GradeSlots? GradeSlotsFor(string grade) =>
        D.GradeSlots.TryGetValue(grade, out var s) ? s : null;

    public static int GradeSlotTotal(string grade)
    {
        if (!D.GradeSlots.TryGetValue(grade, out var s)) return 0;
        return s.Inherent + s.Deco + s.Engr + s.Inscr + s.Extra;
    }

    public static AffixRep? AffixRepFor(string statId) =>
        D.AffixRep.TryGetValue(statId, out var a) ? a : null;
}
